src/common/connection.js

```diff
@@ -176,7 +176,8 @@ class Connection extends EventEmitter {
         this._ws.on('error', error =>
           this.emit('error', 'websocket', error.messsage, error));
         this._ws.on('message', this._onMessage.bind(this));
-        this._onUnexpectedCloseBound = this._onUnexpectedClose.bind(this);
+        this._onUnexpectedCloseBound = this._onUnexpectedClose.bind(this,
+          resolve, reject);
```

Comment:
"close event can came before open" - this is the type of thing that the reader would never guess and should be documented in a comment. Also, it should explain how the `_onUnexpectedCloseBound` is only _temporarily_ assigned here and gets reassigned in `_onOpen`.